### PR TITLE
feat(dtfs2-7075): add trusted user

### PR DESCRIPTION
**Pull request by AlexBramhill**
## Introduction :pencil2:
Adds the trusted user radio button and backend work

## Resolution :heavy_check_mark:
Added toggle on frontend
Updated backend to handle isTrusted

## Miscellaneous :heavy_plus_sign:
As part of the work adding the `isTrusted` role, I've found the need to have partial validation on our payloads.
To fix this I looked at commonising our existing validation of payloads into common lib. I also took the time to convert the file to typescript and utilise zod, which makes our checking more precise.

I've left other non-user validation as is, with the end goal to update this as we address techdebt. I looked into this, and it was requiring too much time commitment.

Added `areAllPropertiesRequired` to allow for checking of updates in our backend validation (such as update users)

image

### CODE REVIEW

File: portal/package.json [hunk 1]
```diff
@@ -55,6 +55,7 @@
     "govuk-frontend": "^4.8.0",
     "http-errors": "^2.0.0",
     "imask": "^6.6.3",
+    "immer": "^10.1.1",
```
Comment (MarRobSoftwire) [inline portal/package.json:58]: Dont think this is used anywhere

File: portal-api/api-tests/v1/users/post-create-user.api-test.js [hunk 2]
```diff
@@ -109,6 +109,7 @@ describe('a user', () => {
                 surname: MOCK_USER.surname,
                 timezone: 'Europe/London',
                 'user-status': STATUS.ACTIVE,
+                isTrusted: false,
```
Comment (MarRobSoftwire) [inline portal-api/api-tests/v1/users/post-create-user.api-test.js:112]:
```suggestion
                isTrusted: MOCK_USER.isTrusted,
```

File: portal-api/src/v1/controllers/banks.controller.js [hunk 3]
```diff
@@ -1,9 +1,8 @@
+const { isVerifiedPayload, BANK } = require('@ukef/dtfs2-common');
```
Comment (MarRobSoftwire) [inline portal-api/src/v1/controllers/banks.controller.js:1]: fwiw importing `BANK` from `@ukef/dtfs2-common` isn't very clear what it's going to be

File: portal-api/src/v1/users/controller.test.js [hunk 4]
```diff
@@ -1,14 +1,26 @@
...
+jest.mock('./sanitizeUserData');
+jest.mock('../email');
+jest.mock('@ukef/dtfs2-common');
+jest.mock('../../crypto/utils');
 const { ObjectId } = require('mongodb');
-const { when } = require('jest-when');
-const { generateMockNoUserLoggedInAuditDatabaseRecord, generateNoUserLoggedInAuditDetails } = require('@ukef/dtfs2-common/change-stream');
+const { when, resetAllWhenMocks } = require('jest-when');
+const {
+  generateMockNoUserLoggedInAuditDatabaseRecord,
```
Comment (MarRobSoftwire) [inline portal-api/src/v1/users/controller.test.js:9]: FYI I've editted this import - one of us will have a conflict

File: trade-finance-manager-api/src/constants/payloads.js [hunk 5]
```diff
@@ -1,49 +0,0 @@
-const BANK = {
-  id: String,
-  name: String,
-  mga: Object,
-  emails: Object,
-  companiesHouseNo: String,
-  partyUrn: String,
-  hasGefAccessOnly: Boolean,
-  paymentOfficerTeam: Object,
-};
```
Comment (MarRobSoftwire) [inline trade-finance-manager-api/src/constants/payloads.js:10]: This BANK object didn't have
```
  utilisationReportPeriodSchedule: 'Object',
  isVisibleInTfmUtilisationReports: 'Boolean',
```
Somewhat surprising that old or new implementation didnt throw any errors

File: package-lock.json [hunk 6]
```diff
@@ -26990,6 +26990,7 @@
         "govuk-frontend": "^4.8.0",
         "http-errors": "^2.0.0",
         "imask": "^6.6.3",
+        "immer": "^10.1.1",
```
Comment (MarRobSoftwire) [inline package-lock.json:26993]: as mentioned, dont think this is used

**Review by MarRobSoftwire — COMMENTED** [review]
Really nice refactoring on the payload validation stuff 😄

File: libs/common/src/helpers/payload-verification/is-verified-payload-by-type.test.ts [hunk 7]
```diff
@@ -18,26 +15,21 @@ describe('isVerifiedPayloadByType', () => {
...
-    },
     {
       payload: { items: [1, 2, 3] },
       template: templates.itemsGeneric,
       description: 'when payload and template have arrays with the template checking for a generic array',
     },
+    {
+      payload: { items: null },
+      template: templates.itemsNull,
+      description: 'when field is an object in template but null in payload',
```
Comment (MarRobSoftwire) [inline libs/common/src/helpers/payload-verification/is-verified-payload-by-type.test.ts:31]: Think this should be 
```suggestion
      description: 'when payload and object are null ',
```

File: libs/common/src/helpers/payload-verification/is-verified-payload-by-type.test.ts [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+
+describe('isVerifiedPayloadByType', () => {
+  const makeIsVerifiedPayloadRequest = isVerifiedPayloadByType;
+
+  const templates = {
+    nameAndAge: { name: 'string', age: 'number' },
+    nameAndAddress: { name: 'string', address: 'object' },
+    nameAndAddressGeneric: { name: 'string', address: 'object' },
+    nameAndAddressIncorrect: { name: 'string', address: 'string' },
+    items: { items: 'object' },
```
[the remark this answers is not in the document]
Comment (MarRobSoftwire) [reply libs/common/src/helpers/payload-verification/is-verified-payload-by-type.test.ts]: Ignore, I'd confused myself

File: libs/common/src/helpers/payload-verification/is-verified-payload-by-type.ts [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+  const templateKeys = Object.keys(template);
+
+  const isEveryTemplateFieldOnPayload = templateKeys.every((propertyKey) => payloadKeys.includes(propertyKey));
+
+  // 2. Ensure no additional properties
+  const isNoAdditionalFieldsOnPayload = !payloadKeys.filter((propertyKey) => !templateKeys.includes(propertyKey)).length;
+
+  // 3. Properties data type validation
+  const isMatchingTypesWithTemplate = payloadKeys.every((key) => {
+    let payloadKeyDataType = typeof payload[key] as string;
```
Comment (MarRobSoftwire) [inline libs/common/src/helpers/payload-verification/is-verified-payload-by-type.ts:22]:
```suggestion
    let payloadKeyDataType = payload[key] === null ? 'null' : typeof payload[key] as string;
```

**Review by MarRobSoftwire — COMMENTED** [review]
A couple minor nits

File: libs/common/src/schemas/audit-database-record.test.ts [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+
+describe('AUDIT_DATABASE_RECORD', () => {
+  const A_VALID_AUDIT_RECORD = generateTfmUserAuditDatabaseRecord(new ObjectId());
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  const { lastUpdatedAt, ...A_AUDIT_RECORD_WITH_MISSING_PARAMETER } = { ...A_VALID_AUDIT_RECORD };
+  const A_AUDIT_RECORD_WITH_INCORRECT_PARAMETER_TYPE = { ...A_VALID_AUDIT_RECORD, lastUpdatedAt: 1 };
+
+  const successTestCases = [
+    {
+      description: 'a valid audit record',
```
Comment (MarRobSoftwire) [inline libs/common/src/schemas/audit-database-record.test.ts:19]: Think these descriptions should be more specific  (e.g. a valid TFM audit record)

**Review by MarRobSoftwire — COMMENTED** [review]
Just a minor nit on test descriptions

**Review by MarRobSoftwire — APPROVED** [review]

**Review by MarRobSoftwire — APPROVED** [review]
LGTM

**Review by ChristianMcc — COMMENTED** [review]
Nice work and payload validation, couple of small improvements to make!

**Review by ChristianMcc — APPROVED** [review]

**Pull request merged**